src/memory.rs

```diff
@@ -446,9 +452,24 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
 /// Reading and writing
 impl<'a, 'tcx> Memory<'a, 'tcx> {
     pub fn freeze(&mut self, alloc_id: AllocId) -> EvalResult<'tcx, ()> {
+        // FIXME: the comment is wrong and do we really still need this check?
         // It's not possible to freeze the zero-sized allocation, because it doesn't exist.
         if alloc_id != ZST_ALLOC_ID {
-            self.get_mut(alloc_id)?.immutable = true;
+            // do not use `self.get_mut(alloc_id)` here, because we might have already frozen a
+            // sub-element or have circular pointers (e.g. `Rc`-cycles)
+            let allocs: Vec<_> = match self.alloc_map.get_mut(&alloc_id) {
+                Some(ref mut alloc) if !alloc.immutable => {
+                    alloc.immutable = true;
+                    alloc.relocations.values().cloned().collect()
```

Comment:
Can we make this work without allocating a `Vec`? Kill the `.collect()`, let the `allocs` var infer to the iterator type, and make the `Vec::new()` branches `return Ok(())`.

Or at that point, move the `for alloc in allocs` loop into this match arm, and let the `Vec::new()` arms fall through to the regular `Ok(())` return.